pubsub-mapped-api/src/test/java/com/google/pubsub/clients/consumer/KafkaConsumerTest.java

```diff
@@ -0,0 +1,300 @@
+package com.google.pubsub.clients.consumer;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.powermock.api.mockito.PowerMockito.when;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import com.google.pubsub.common.ChannelUtil;
+import com.google.pubsub.common.StubCreator;
+import com.google.pubsub.kafkastubs.common.KafkaException;
+import com.google.pubsub.kafkastubs.common.serialization.IntegerSerializer;
+import com.google.pubsub.kafkastubs.common.serialization.StringSerializer;
+import com.google.pubsub.kafkastubs.consumer.ConsumerRecord;
+import com.google.pubsub.kafkastubs.consumer.ConsumerRecords;
+import com.google.pubsub.v1.ListTopicsResponse;
+import com.google.pubsub.v1.PublisherGrpc;
+import com.google.pubsub.v1.PublisherGrpc.PublisherBlockingStub;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.ReceivedMessage;
+import com.google.pubsub.v1.SubscriberGrpc;
+import com.google.pubsub.v1.SubscriberGrpc.SubscriberBlockingStub;
+import com.google.pubsub.v1.SubscriberGrpc.SubscriberFutureStub;
+import com.google.pubsub.v1.Subscription;
+import com.google.pubsub.v1.Topic;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.core.classloader.annotations.SuppressStaticInitializationFor;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({SubscriberGrpc.class, PublisherGrpc.class, Futures.class})
```

Comment:
Don't do this. gRPC does not support mocking in many scenarios, and we have purposefully made classes final to prevent people from mocking them. If you mock things to this extent two things happen 1) your tests become change detectors that don't find actual bugs (they don't help you while harming productivity) and 2) upgrades to gRPC can trivially break the tests.

Instead of mocking, you should use the actual implementations and fake out what you need. Mocking should be the last tool you choose to use instead of the first. (Mocking does work well for recording notifications, when the mock doesn't have to do anything when called.)

Instead of mocking the stub, fake or mock the service implementation and use the inprocess transport to use a real stub to communicate with the fake service. GrpcServerRule does most of this for you. Take a look at the client tests for [helloworld and routeguide examples](https://github.com/grpc/grpc-java/tree/v1.5.0/examples/src/test/java/io/grpc/examples).